### PR TITLE
import/add-index: fix race and performance regression

### LINKED ISSUE #44584: DDL: The performance drop 70%
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. Deploy a TiDB cluster with 1 TiDB and 6 TiKV.
2. Restore the data from the backup of [ossinsight](https://ossinsight.io/docs/workshop/ossinsight-lite/introduction) system
3. `set global tidb_ddl_enable_fast_reorg=ON;`
4. `alter table github_events add index test_idx(repo_id,type,action,closed_at,actor_login);`

### 2. What did you expect to see? (Required)
No performance regression

### 3. What did you see instead (Required)
image

### 4. What is your TiDB version? (Required)

good commit: be8d86504402cf6e9648fa4972d29e6ac3f9e9a8
bad commit: 6fbf29981ed5facc1f47fcf041386254bb5e88d0

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve?

Issue Number: close #44584 

Problem Summary:
```
WARNING: DATA RACE
Write at 0x00c006882480 by goroutine 304253:
  github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Writer).appendRowsSorted()
      br/pkg/lightning/backend/local/engine.go:1028 +0xa5
  github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Writer).AppendRows()
      br/pkg/lightning/backend/local/engine.go:1124 +0x3e9
  github.com/pingcap/tidb/executor/importer.(*chunkProcessor).deliverLoop.func1()
      executor/importer/chunk_process.go:253 +0x1cd
  github.com/pingcap/tidb/executor/importer.(*chunkProcessor).deliverLoop()
      executor/importer/chunk_process.go:266 +0x578
  github.com/pingcap/tidb/executor/importer.(*chunkProcessor).process.func2()
      executor/importer/chunk_process.go:131 +0x4c
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      external/org_golang_x_sync/errgroup/errgroup.go:75 +0x82

Previous read at 0x00c006882480 by goroutine 304138:
  github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Writer).EstimatedSize()
      br/pkg/lightning/backend/local/engine.go:1160 +0x64
  github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Engine).getEngineFileSize.func1()
      br/pkg/lightning/backend/local/engine.go:489 +0x7b
  sync.(*Map).Range()
      GOROOT/src/sync/map.go:476 +0x1db
  github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Engine).getEngineFileSize()
      br/pkg/lightning/backend/local/engine.go:487 +0x165
  github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).EngineFileSizes.func1()
      br/pkg/lightning/backend/local/local.go:1744 +0x75
  sync.(*Map).Range()
      GOROOT/src/sync/map.go:476 +0x1db
  github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).EngineFileSizes()
      br/pkg/lightning/backend/local/local.go:1742 +0x64
  github.com/pingcap/tidb/br/pkg/lightning/backend/local.CheckDiskQuota()
      br/pkg/lightning/backend/local/disk_quota.go:40 +0x61
  github.com/pingcap/tidb/executor/importer.(*TableImporter).CheckDiskQuota()
      executor/importer/table_import.go:492 +0x1f6
  github.com/pingcap/tidb/disttask/importinto.(*importStepScheduler).InitSubtaskExecEnv.func1()
      disttask/importinto/scheduler.go:87 +0xbb

Goroutine 304253 (running) created at:
  golang.org/x/sync/errgroup.(*Group).Go()
      external/org_golang_x_sync/errgroup/errgroup.go:72 +0x12e
  github.com/pingcap/tidb/executor/importer.(*chunkProcessor).process()
      executor/importer/chunk_process.go:130 +0x1f9
  github.com/pingcap/tidb/executor/importer.ProcessChunk()
      executor/importer/engine_process.go:100 +0xc69
  github.com/pingcap/tidb/disttask/importinto.(*ImportMinimalTaskExecutor).Run()
      disttask/importinto/subtask_executor.go:65 +0x70e
  github.com/pingcap/tidb/disttask/framework/scheduler.(*InternalSchedulerImpl).runMinimalTask()
      disttask/framework/scheduler/scheduler.go:292 +0xac7
  github.com/pingcap/tidb/disttask/framework/scheduler.(*InternalSchedulerImpl).runSubtask.func1()
      disttask/framework/scheduler/scheduler.go:219 +0x21a
  github.com/pingcap/tidb/resourcemanager/pool/spool.runTask()
      resourcemanager/pool/spool/spool.go:229 +0xfe
  github.com/pingcap/tidb/resourcemanager/pool/spool.(*Pool).RunWithConcurrency.func1()
      resourcemanager/pool/spool/spool.go:166 +0x30
  github.com/pingcap/tidb/resourcemanager/pool/spool.(*Pool).run.func1()
      resourcemanager/pool/spool/spool.go:145 +0x72

Goroutine 304138 (running) created at:
  github.com/pingcap/tidb/disttask/importinto.(*importStepScheduler).InitSubtaskExecEnv()
      disttask/importinto/scheduler.go:85 +0xa19
  github.com/pingcap/tidb/disttask/framework/scheduler.(*InternalSchedulerImpl).run()
      disttask/framework/scheduler/scheduler.go:127 +0x473
  github.com/pingcap/tidb/disttask/framework/scheduler.(*InternalSchedulerImpl).Run()
      disttask/framework/scheduler/scheduler.go:104 +0x4e
  github.com/pingcap/tidb/disttask/framework/scheduler.(*Manager).onRunnableTask()
      disttask/framework/scheduler/manager.go:293 +0xd8f
  github.com/pingcap/tidb/disttask/framework/scheduler.(*Manager).onRunnableTasks.func1()
      disttask/framework/scheduler/manager.go:204 +0x90
  github.com/pingcap/tidb/resourcemanager/pool/spool.(*Pool).run.func1()
      resourcemanager/pool/spool/spool.go:145 +0x72
==================
==================
WARNING: DATA RACE
Write at 0x00c0068824d8 by goroutine 304253:
  github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Writer).appendRowsSorted()
      br/pkg/lightning/backend/local/engine.go:1035 +0x2c4
  github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Writer).AppendRows()
      br/pkg/lightning/backend/local/engine.go:1124 +0x3e9
  github.com/pingcap/tidb/executor/importer.(*chunkProcessor).deliverLoop.func1()
      executor/importer/chunk_process.go:253 +0x1cd
  github.com/pingcap/tidb/executor/importer.(*chunkProcessor).deliverLoop()
      executor/importer/chunk_process.go:266 +0x578
  github.com/pingcap/tidb/executor/importer.(*chunkProcessor).process.func2()
      executor/importer/chunk_process.go:131 +0x4c
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      external/org_golang_x_sync/errgroup/errgroup.go:75 +0x82

Previous read at 0x00c0068824d8 by goroutine 304138:
  github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Writer).EstimatedSize()
      br/pkg/lightning/backend/local/engine.go:1165 +0x88
  github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Engine).getEngineFileSize.func1()
      br/pkg/lightning/backend/local/engine.go:489 +0x7b
  sync.(*Map).Range()
      GOROOT/src/sync/map.go:476 +0x1db
  github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Engine).getEngineFileSize()
      br/pkg/lightning/backend/local/engine.go:487 +0x165
  github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).EngineFileSizes.func1()
      br/pkg/lightning/backend/local/local.go:1744 +0x75
  sync.(*Map).Range()
      GOROOT/src/sync/map.go:476 +0x1db
  github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).EngineFileSizes()
      br/pkg/lightning/backend/local/local.go:1742 +0x64
  github.com/pingcap/tidb/br/pkg/lightning/backend/local.CheckDiskQuota()
      br/pkg/lightning/backend/local/disk_quota.go:40 +0x61
  github.com/pingcap/tidb/executor/importer.(*TableImporter).CheckDiskQuota()
      executor/importer/table_import.go:492 +0x1f6
  github.com/pingcap/tidb/disttask/importinto.(*importStepScheduler).InitSubtaskExecEnv.func1()
      disttask/importinto/scheduler.go:87 +0xbb

Goroutine 304253 (running) created at:
  golang.org/x/sync/errgroup.(*Group).Go()
      external/org_golang_x_sync/errgroup/errgroup.go:72 +0x12e
  github.com/pingcap/tidb/executor/importer.(*chunkProcessor).process()
      executor/importer/chunk_process.go:130 +0x1f9
  github.com/pingcap/tidb/executor/importer.ProcessChunk()
      executor/importer/engine_process.go:100 +0xc69
  github.com/pingcap/tidb/disttask/importinto.(*ImportMinimalTaskExecutor).Run()
      disttask/importinto/subtask_executor.go:65 +0x70e
  github.com/pingcap/tidb/disttask/framework/scheduler.(*InternalSchedulerImpl).runMinimalTask()
      disttask/framework/scheduler/scheduler.go:292 +0xac7
  github.com/pingcap/tidb/disttask/framework/scheduler.(*InternalSchedulerImpl).runSubtask.func1()
      disttask/framework/scheduler/scheduler.go:219 +0x21a
  github.com/pingcap/tidb/resourcemanager/pool/spool.runTask()
      resourcemanager/pool/spool/spool.go:229 +0xfe
  github.com/pingcap/tidb/resourcemanager/pool/spool.(*Pool).RunWithConcurrency.func1()
      resourcemanager/pool/spool/spool.go:166 +0x30
  github.com/pingcap/tidb/resourcemanager/pool/spool.(*Pool).run.func1()
      resourcemanager/pool/spool/spool.go:145 +0x72

Goroutine 304138 (running) created at:
  github.com/pingcap/tidb/disttask/importinto.(*importStepScheduler).InitSubtaskExecEnv()
      disttask/importinto/scheduler.go:85 +0xa19
  github.com/pingcap/tidb/disttask/framework/scheduler.(*InternalSchedulerImpl).run()
      disttask/framework/scheduler/scheduler.go:127 +0x473
  github.com/pingcap/tidb/disttask/framework/scheduler.(*InternalSchedulerImpl).Run()
      disttask/framework/scheduler/scheduler.go:104 +0x4e
  github.com/pingcap/tidb/disttask/framework/scheduler.(*Manager).onRunnableTask()
      disttask/framework/scheduler/manager.go:293 +0xd8f
  github.com/pingcap/tidb/disttask/framework/scheduler.(*Manager).onRunnableTasks.func1()
      disttask/framework/scheduler/manager.go:204 +0x90
  github.com/pingcap/tidb/resourcemanager/pool/spool.(*Pool).run.func1()
      resourcemanager/pool/spool/spool.go:145 +0x72
```
### What is changed and how it works?
- remove lock when get engine size
- add writerSize to fix data race
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/local/engine.go [hunk 1]
```diff
@@ -1009,7 +1000,8 @@ type Writer struct {
 	// if the KVs are append in order, we can directly write the into SST file,
 	// else we must first store them in writeBatch and then batch flush into SST file.
 	isKVSorted bool
-	writer     *sstWriter
+	writer     atomic.Pointer[sstWriter]
+	writerSize atomic.Uint64
```
Comment (D3Hunter) [inline br/pkg/lightning/backend/local/engine.go:1004]: we can call `EstimatedSize`(quite cheap) directly, no need to store this.

Comment (lance6716) [reply br/pkg/lightning/backend/local/engine.go:1004]: I remember the problem is sstable.Writer's EstimatedSize is not concurrent safe.

File: br/pkg/lightning/backend/local/engine.go [hunk 2]
```diff
@@ -1158,6 +1157,16 @@ func (w *Writer) flush(ctx context.Context) error {
 	return nil
 }
 
+// EstimatedSize returns the estimated size of the SST file.
+func (w *Writer) EstimatedSize() uint64 {
+	if w.writer.Load() != nil {
+		return w.writerSize.Load()
```
Comment (lance6716) [inline br/pkg/lightning/backend/local/engine.go:1163]: the read of `writer` and `writerSize` is not atomic. Can we check writerSize.Load() != 0 instead?

**Review by D3Hunter — APPROVED** [review]

File: br/pkg/lightning/backend/local/engine.go (hunk 1 above)
Comment (D3Hunter) [reply br/pkg/lightning/backend/local/engine.go:1004]: lgtm

**Pull request merged**